traefik/traefik_utils.py

```diff
@@ -0,0 +1,149 @@
+import time
+import socket
+import sys
+import requests
+from os.path import abspath, dirname, join
+from subprocess import Popen
+
+_ports = {
+    "traefik": 8000,
+    "default_backend": 9000,
+    "first_backend": 9090,
+    "second_backend": 9099,
+}
+
+
+def get_port(service_name):
+    return _ports[service_name]
+
+
+def is_open(ip, port):
+    timeout = 1
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    s.settimeout(timeout)
+    try:
+        s.connect((ip, port))
+        s.shutdown(socket.SHUT_RDWR)
+        return True
+    except:
+        return False
+    finally:
+        s.close()
+
+
+def check_host_up(ip, port):
+    """ Allow the service up to 1 sec to open connection on the
+    designated port """
+    up = False
+    retry = 20  # iterations
+    delay = 0.05  # 50 ms
+
+    for i in range(retry):
+        if is_open(ip, port):
+            up = True
+            break
+        else:
+            time.sleep(delay)
+    return up
+
+
+def traefik_routes_to_correct_backend(path, expected_port):
+    """ Check if traefik followed the configuration and routed the
+    request to the right backend """
+    base_url = "http://localhost:" + str(get_port("traefik"))
+    resp = requests.get(base_url + path)
+    assert int(resp.text) == expected_port
+
+
+def check_traefik_etcd_conf_ready():
+    base_url = "http://localhost:" + str(get_port("traefik"))
+    """ Allow traefik up to 3 sec to load its configuration from the
+    etcd cluster """
+    timeout = time.time() + 3
+    ready = False
+    while ready is False:
+        resp = requests.get(base_url + "/api/providers/etcdv3")
+        ready = resp.status_code == 200
```

Comment:
This is going to try *really* fast. If we add a backoff sleep, we can get a little better behavior. Additionally, since things can sometimes be super slow on Travis, we can pick a more conservative timeout of 10-15 seconds:

```python
timeout = time.time() + 10
ready = False
t = 0.1
while not ready and time.time() < timeout:
    ...
    if not ready:
        t = min(2, t * 2)
        time.sleep(t)
assert ready # check that we got here because we are ready